### PR TITLE
Convert "Grass7AlgorithmProvider.settingName" to a class attribute.

### LINKED ISSUE #30241: Traceback when closing QGIS after installing a custom GRASS Processing Provider
Checklist before submitting

- [v] Search through existing issue reports and gis.stackexchange.com to check whether the issue already exists

**Describe the bug**

After opening a GRASS mapset, closing QGIS throws a Python Traceback in a pop-up window.

**How to Reproduce**

1. Open QGIS with GRASS 
2. Open a GRASS mapset. I used the Alaska demo dataset
3. [optionally] Close the mapset. You get the traceback even if you omit this step.
4. Close QGIS
5. The traceback pops up:

```
An error has occurred while executing Python code: 

KeyError: 'ACTIVATE_GRASS7' 
Traceback (most recent call last):
  File "/opt/qgis-git/share/qgis/python/plugins/processing/algs/grass7/Grass7AlgorithmProvider.py", line 88, in unload
    ProcessingConfig.removeSetting('ACTIVATE_GRASS7')
  File "/opt/qgis-git/share/qgis/python/plugins/processing/core/ProcessingConfig.py", line 157, in removeSetting
    del ProcessingConfig.settings[name]
KeyError: 'ACTIVATE_GRASS7'


Python version: 3.7.3 (default, Mar 26 2019, 21:43:19) [GCC 8.2.1 20181127] 
QGIS version: 3.7.0-Master Master, 52d22b8050 

Python Path:
/opt/qgis-git/share/qgis/python
/home/feanor/.local/share/QGIS/QGIS3/profiles/default/python
/home/feanor/.local/share/QGIS/QGIS3/profiles/default/python/plugins
/opt/qgis-git/share/qgis/python/plugins
/usr/lib/python37.zip
/usr/lib/python3.7
/usr/lib/python3.7/lib-dynload
/home/feanor/.local/lib/python3.7/site-packages
/usr/lib/python3.7/site-packages
/home/feanor/.local/share/QGIS/QGIS3/profiles/default/python
```

https://i.imgur.com/vY3bqpN.png

**QGIS and OS versions**

QGIS version | 3.7.0-Master | QGIS code revision | 52d22b8050
-- | -- | -- | --
Compiled against Qt | 5.12.3 | Running against Qt | 5.12.3
Compiled against GDAL/OGR | 2.4.1 | Running against GDAL/OGR | 2.4.1
Compiled against GEOS | 3.7.1-CAPI-1.11.1 | Running against GEOS | 3.7.1-CAPI-1.11.1 27a5e771
PostgreSQL Client Version | 11.3 | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.1
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | Arch Linux |   |  

**More context**

I did debug a little bit and it seems that `Grass7AlgorithmProvider.unload()` is being called twice. The first time `ACTIVATE_GRASS7` gets deleted, but the second time a `KeyError` is being raised.

### PR DESCRIPTION
**Pull request by pmav99**
Subclasses of `Grass7AlgorithmProvider` were using the same setting name as
the parent class (i.e. 'ACTIVATE_GRASS7'). This was causing a `KeyError`
when QGIS was being closed.

With this commit we convert the setting name key to a class attribute.
This way we reduce code maintenance and we reduce the number of methods
that `Grass7AlgorithmProvider` subclasses need to override.

Fixes #30241
Continues #9202

### CODE REVIEW

**Comment by nyalldawson** [comment]
This looks good to me.

One thing I think may be worth considering before merge: is it actually needed to have a setting to enable your derived provider? In my view this setting is completely redundant for plugin based providers, because a user will either not install the plugin of they don't want the provider, or they can just disable the plugin itself. It seems totally pointless to me for someone to install a plugin and then disable the plugin's functionality via a switch like this.

So maybe this could be tweaked to avoid the setting if the key is empty?

**Comment by pmav99** [comment]
Thanks for the review Nyall.

My knowledge of QGIS internals is rather limited so please bear with me. 

From a UI/UX point of view, I don't think that it is necessary to actually have such a switch. To be honest, I didn't even know what `ACTIVATE_*` was being used for... :) 

That being said though, if we want to avoid changing `Grass7AlgorithmProvider` interface + to keep subclasses implementations simple then, not setting the key requires some refactoring. E.g. a `Grass7AlgorithmProviderBase` parent class to inherit from:

``` python

# (the actual implementation might be cleaner if `abc.abstractproperty` is being used
# but for the time being let's avoid metaclasses etc)

# The base/abstract implementation of the core Grass7 provider plugin
class Grass7AlgorithmProviderBase(QgsProcessingProvider):

    @property
    def descriptionFolder(self):
        raise NotImplementedError("You need to override descriptionFolder in your subclasses")

    def __init__(self):
        super().__init__()
        self.algs = []

    def load(self):
        ProcessingConfig.settingIcons[self.name()] = self.icon()
        ProcessingConfig.readSettings()
        self.refreshAlgorithms()
        return True

    def unload(self):
        pass

    def is_active(self):
        return None  # XXX or maybe False

    def set_active(self, active):
        pass

    # the rest of the common methods/attributes
    # ...

# The concrete implementation of the core Grass7 provider plugin
class Grass7AlgorithmProvider(Grass7AlgorithmProviderBase):

    @property
    def descriptionFolder(self):
        return Grass7Utils.grassDescriptionPath()

    def load(self):
        super().load()
        ProcessingConfig.addSetting(Setting(self.name(), self.settingName,
                                            self.tr('Activate'), True))
        if isMac():
            ProcessingConfig.addSetting(Setting(
                self.name(),
                Grass7Utils.GRASS_FOLDER, self.tr('GRASS7 folder'),
                Grass7Utils.grassPath(), valuetype=Setting.FOLDER))
        ProcessingConfig.addSetting(Setting(
            self.name(),
            Grass7Utils.GRASS_LOG_COMMANDS,
            self.tr('Log execution commands'), False))
        ProcessingConfig.addSetting(Setting(
            self.name(),
            Grass7Utils.GRASS_LOG_CONSOLE,
            self.tr('Log console output'), False))
        ProcessingConfig.addSetting(Setting(
            self.name(),
            Grass7Utils.GRASS_HELP_PATH,
            self.tr('Location of GRASS docs'),
            Grass7Utils.grassHelpPath()))
        # Add a setting for using v.external instead of v.in.ogr
        # But set it to False by default because some algorithms
        # can't be used with external data (need a solid v.in.ogr).
        ProcessingConfig.addSetting(Setting(
            self.name(),
            Grass7Utils.GRASS_USE_VEXTERNAL,
            self.tr('For vector layers, use v.external (faster) instead of v.in.ogr'),
            False))
        ProcessingConfig.readSettings()
        self.refreshAlgorithms()
        return True

    def unload(self):
        super().unload()
        ProcessingConfig.removeSetting(self.settingName)
        if isMac():
            ProcessingConfig.removeSetting(Grass7Utils.GRASS_FOLDER)
        ProcessingConfig.removeSetting(Grass7Utils.GRASS_LOG_COMMANDS)
        ProcessingConfig.removeSetting(Grass7Utils.GRASS_LOG_CONSOLE)
        ProcessingConfig.removeSetting(Grass7Utils.GRASS_HELP_PATH)
        ProcessingConfig.removeSetting(Grass7Utils.GRASS_USE_VEXTERNAL)

    def isActive(self):
        return ProcessingConfig.getSetting(self.settingName)

    def setActive(self, active):
        ProcessingConfig.setSettingValue(self.settingName, active)


# Implementation of a 3rd party Grass based Provider
class CustomGrassBasedProvider(Grass7AlgorithmProviderBase):

    @property
    def descriptionFolder(self):
        return os.path.join(os.path.dirname(__file__), 'description')
```
Is something like this what you have in mind?

**Comment by nyalldawson** [comment]
Yes, looks good! So we'd just need to make sure that the plugin based subclasses inherit the other settings, eg the path related ones, but ignore the activation setting.

**Comment by alexbruy** [comment]
Maybe I missed something, but some time ago we agreed that `Grass7AlgorithmProvider` and all other similar classes, including algorithm classes like `GdalAlgorithm` should be considired as private API and using them by 3rd party plugins should be discouraged. Or at least plugin authors should be ready that API and functionality may change without any notification.

**Comment by pmav99** [comment]
@alexbruy The approach we used for the QGIS plugin we developed was implemented after @nyalldawson 's suggestion ([source](https://lists.osgeo.org/pipermail/qgis-developer/2019-February/056155.html)). I am not familiar with the discussion you mention (link please? :) ), nevertheless, since GRASS has several hundreds addons, having a way to expose them as QGIS plugins is something that IMHV is not meritless. 

BTW, I am not claiming that using `Grass7AlrorithmProvider` is necessarily the optimal way but at the same time it feels like a reasonable way. As far as API stability goes, I understand why you want to avoid a puclic API and, at least for now, I am perfectly fine with no stability guarantees.

**Comment by pmav99** [comment]
I forced push two new commits: 
- The first one just removes the hard-coded "ACTIVATE_GRASS7" key.
- The second one makes setting the key optional.
- I also removed the comment that provided information WRT subclassing `Grass7AlgorithmProvider`

If someone truly wants to subclass the API he should still be able to do it like this:

``` python
class CustomGrassBasedProvider(Grass7AlgorithmProvider):

    descriptionFolder = os.path.join(os.path.dirname(__file__), 'description')
    activateSetting = None

    # define the rest of the methods that are needed (e.g. name(), id() etc)
    # ...
```

**Comment by nyalldawson** [comment]
@alexbruy I had the same initial concerns as you, but @pmav99 has made it clear that they don't expect stable API here. So I'm fine with modifications to the provider to allow this use case, as long as they don't impact on the core grass provider functionality (which this doesn't).

File: python/plugins/processing/algs/grass7/Grass7AlgorithmProvider.py [hunk 1]
```diff
@@ -92,10 +93,12 @@ def unload(self):
         ProcessingConfig.removeSetting(Grass7Utils.GRASS_USE_VEXTERNAL)
 
     def isActive(self):
-        return ProcessingConfig.getSetting('ACTIVATE_GRASS7')
+        if self.activateSetting:
+            return ProcessingConfig.getSetting(self.activateSetting)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/grass7/Grass7AlgorithmProvider.py]:
```suggestion
            return ProcessingConfig.getSetting(self.activateSetting)

        return True
```

**Comment by nyalldawson** [comment]
Sorry @pmav99  - I missed the change you fixed here (can you ping after force pushing a change like this? Github doesn't notify automatically in this situation)

**Comment by pmav99** [comment]
@nyalldawson, sure no problem.

BTW, should I backport this to LTR, too?

**Comment by nyalldawson** [comment]
@pmav99 up to you, but I think it's safe to do.

**Pull request merged**